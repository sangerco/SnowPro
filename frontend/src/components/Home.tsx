import React from 'react';
import './Home.css';
import { useAuth } from './AuthProvider';
import { fetchUserData } from '../redux/actions/userActions';


const Home: React.FC = () => {
    return <p>home</p>
}

export default Home;