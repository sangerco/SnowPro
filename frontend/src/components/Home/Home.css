.home-container {
    background-image: url(https://drive.google.com/file/d/1yXhFch_OTpTjW3QCu43jChlOTlQp3Z5x/view?usp=sharing);
    background-size: cover;
    background-position: center;
    background-position: no-repeat;
    background-attachment: fixed;
    width: 100%;
    height: 100vh;
}