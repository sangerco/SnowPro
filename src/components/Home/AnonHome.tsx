import React, { useEffect } from "react";

const AnonHome: React.FC = () => {
  return (
    <div>
      <p>Anon Home</p>
    </div>
  );
};

export default AnonHome;
