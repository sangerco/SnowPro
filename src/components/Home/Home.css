body {
    background-image: url('../../media/sebastian-staines-xZIFEPtRsRI-unsplash.jpg') !important;
    background-size: cover !important;
    background-repeat: no-repeat !important;
    background-attachment: fixed !important;
}

#review-card {
    margin-top: 10px;
}