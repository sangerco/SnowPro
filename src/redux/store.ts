import { configureStore, combineReducers } from "@reduxjs/toolkit";
import userReducer from "./slices/userSlice";
import authReducer from "./slices/authSlice";
import messageReducer from "./slices/messageSlice";
import messageReplyReducer from "./slices/messageReplySlice";
import reviewReducer from "./slices/reviewSlice";
import reviewReplyReducer from "./slices/reviewReplySlice";
import tagReducer from "./slices/tagSlice";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";

const persistConfig = {
  key: "root",
  storage,
  blacklist: ["messages", "messageReplies", "reviews", "reviewReplies"],
};

const rootReducer = combineReducers({
  users: userReducer,
  auth: authReducer,
  messages: messageReducer,
  messageReplies: messageReplyReducer,
  reviews: reviewReducer,
  reviewReplies: reviewReplyReducer,
  tags: tagReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
