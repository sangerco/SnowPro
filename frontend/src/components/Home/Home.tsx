import React, { useEffect } from 'react';
import './Home.css';
import { useAuth } from '../AuthProvider';
import { RootState } from '../../redux/store'
import { fetchUserData } from '../../redux/actions/userActions';
import { useSelector } from 'react-redux';

interface UserData {
    id: string;
    username: string;
    firstName: string;
    lastName: string;
    email: string;
    avatar: string;
    bio: string;
    videos: string[];
    photos: string[];
    favMountains: string[];
}

interface HomePageProps {
    user: UserData | null;
    fetchUserData: ( username: string) => void;
}

const Home: React.FC<HomePageProps> = ({ user, fetchUserData }) => {
    const { username } = useAuth();

    const isAuthenticated = useSelector((state: RootState) => state.login.token)
    
    useEffect(() => {
        if(isAuthenticated) { 
            fetchUserData(username as string);
            }
    }, [isAuthenticated, fetchUserData]);

    if(isAuthenticated) {
        return <p>logged in!</p>
    } else {
        return <p>not logged in!</p>
    }
}

export default Home;