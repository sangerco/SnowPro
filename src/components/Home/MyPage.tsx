import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  Dimmer,
  Loader,
  Grid,
  Container,
  Header,
  Image,
  Modal,
  Button,
  Icon,
  Embed,
  Card,
  Divider,
} from "semantic-ui-react";
import { Link } from "react-router-dom";
import { RootState, AppDispatch } from "../../redux/store";
import { fetchOneUser, deleteUser } from "../../redux/slices/userSlice";
import PhotoForm from "../Media/PhotoForm";
import VideoForm from "../Media/VideoForm";
import FavMountain from "../SkiAreas/FavMountain";

const MyPage = () => {
  const auth = useSelector((state: RootState) => state.auth);
  const dispatch = useDispatch<AppDispatch>();
  const username = auth.data?.username;

  const [showDeleteModal, setShowDeleteModal] = useState(false);
  const [showNewPhotoForm, setShowNewPhotoForm] = useState(false);
  const [showNewVideoForm, setShowNewVideoForm] = useState(false);

  useEffect(() => {
    if (username) {
      dispatch(fetchOneUser(username));
    }
  }, [dispatch, username]);

  const user = useSelector((state: RootState) => state.users.user);

  const handleShowDeleteModal = () => {
    setShowDeleteModal(true);
  };

  const handleDeleteUser = () => {
    if (username) {
      dispatch(deleteUser(username));
    }
  };

  let videoLinks: string[] = [];

  if (user && user.videos && user.videos.length > 0) {
    user.videos.forEach((link) => {
      let splitLink = link.split("v=");
      let id = splitLink[1];
      let finalId = id.includes("&") ? id.split("&")[0] : id;
      videoLinks.push(finalId);
    });
  }

  if (auth.loading) {
    return (
      <div>
        <Dimmer active>
          <Loader>Loading...</Loader>
        </Dimmer>
      </div>
    );
  }

  if (user) {
    return (
      <div>
        <Container>
          <Grid relaxed>
            <Grid.Row>
              <Grid.Column width={4}>
                <Image src={user.avatar} size="small" />
              </Grid.Column>
              <Grid.Column width={8}>
                <Header as="h2" textAlign="right">
                  {user.username}
                </Header>
                <Header.Subheader>
                  {user.firstName && user.firstName !== ""
                    ? user.firstName
                    : ""}{" "}
                  {user.lastName && user.lastName !== "" ? user.lastName : ""}
                </Header.Subheader>
                <Link to={`/users/${user.username}/update`}>
                  <Icon name="edit" style={{ cursor: "pointer" }} />
                </Link>
                <Icon
                  name="trash"
                  style={{ cursor: "pointer" }}
                  onClick={handleShowDeleteModal}
                />
              </Grid.Column>
              <Grid.Column width={4}></Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column width={2}></Grid.Column>
              <Grid.Column width={6}>
                <p>{user.bio}</p>
              </Grid.Column>
              <Grid.Column width={8}>
                <Header as="h3">Favorite Mountains</Header>
                <Divider />
                {user.favMountains &&
                username &&
                user.favMountains.length > 0 ? (
                  user.favMountains.map((fm) => (
                    <Card.Group key={fm}>
                      <FavMountain key={fm} slug={fm} username={username} />
                    </Card.Group>
                  ))
                ) : (
                  <p>This user has no favorited mountains yet!</p>
                )}
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column width={2}></Grid.Column>
              <Grid.Column width={12}>
                {user.photos && user.photos.length > 0
                  ? user.photos.map((link) => (
                      <Image
                        src={link}
                        size="small"
                        as={Link}
                        to={`/users/${user.username}/photos/`}
                      />
                    ))
                  : null}
                <Button
                  size="small"
                  color="green"
                  onClick={() => setShowNewPhotoForm(true)}>
                  Add New Photo?
                </Button>
                {showNewPhotoForm && <PhotoForm />}
              </Grid.Column>
              <Grid.Column width={2}></Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column width={2}></Grid.Column>
              <Grid.Column width={12}>
                {videoLinks.length > 0
                  ? videoLinks.map((link) => (
                      <div style={{ marginTop: "20px" }}>
                        <Embed id={link} source="youtube" size="small" />
                      </div>
                    ))
                  : null}
                <Button
                  style={{ marginTop: "20px" }}
                  color="green"
                  size="small"
                  onClick={() => setShowNewVideoForm(true)}>
                  Add New Video?
                </Button>
                {showNewVideoForm && <VideoForm />}
              </Grid.Column>
              <Grid.Column width={2}></Grid.Column>
            </Grid.Row>
          </Grid>
        </Container>
        <Modal open={showDeleteModal} onClose={() => setShowDeleteModal(false)}>
          <Modal.Content>
            Are You Sure You Want To Delete Your Profile?
          </Modal.Content>
          <Modal.Actions>
            <Button negative onClick={handleDeleteUser}>
              Yes
            </Button>
            <Button onClick={() => setShowDeleteModal(false)}>Cancel</Button>
          </Modal.Actions>
        </Modal>
      </div>
    );
  }

  if (auth.error) {
    return <div>Error: {auth.error}</div>;
  }

  return null;
};

export default MyPage;
