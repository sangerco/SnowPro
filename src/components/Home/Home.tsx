import React, { useEffect, useState } from "react";
import MyPage from "./MyPage";
import { fetchSkiAreas } from "../../redux/slices/skiAreaSlice";
import { fetchAllReviews } from "../../redux/slices/reviewSlice";
import AnonHome from "./AnonHome";
import "./Home.css";
import {
  Card,
  Dimmer,
  Divider,
  Grid,
  Header,
  Loader,
  Pagination,
  PaginationProps,
  Rating,
  Segment,
} from "semantic-ui-react";
import { useDispatch, useSelector } from "react-redux";
import { AppDispatch, RootState } from "../../redux/store";
import { Link } from "react-router-dom";
import Inbox from "../Messages/Inbox";

const Home: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const isAuthenticated = useSelector(
    (state: RootState) => state.auth.isAuthenticated
  );
  const username = useSelector((state: RootState) => state.auth.data?.username);

  useEffect(() => {
    dispatch(fetchSkiAreas());
  }, [dispatch]);

  const skiAreaState = useSelector((state: RootState) => state.skiAreas);
  const skiAreas = skiAreaState.skiAreas;

  useEffect(() => {
    dispatch(fetchAllReviews());
  }, [dispatch]);

  const reviewState = useSelector((state: RootState) => state.reviews);
  const reviews = reviewState.reviews;

  let totalReviews;
  let paginatedReviews;

  const itemsPerPage = 3;
  const [activePage, setActivePage] = useState(1);
  reviews ? (totalReviews = reviews.length) : (totalReviews = 1);
  const totalPages = Math.ceil(totalReviews / itemsPerPage);
  const startIndex = (activePage - 1) * itemsPerPage;
  const endIndex = Math.min(startIndex + itemsPerPage, totalReviews);
  reviews
    ? (paginatedReviews = reviews.slice(startIndex, endIndex))
    : (paginatedReviews = 1);

  const handlePaginationChange = (
    e: React.MouseEvent,
    data: PaginationProps
  ) => {
    setActivePage(data.activePage as number);
  };

  const formatDate = (date: Date) => {
    const newDate = new Date(date);
    const options: Intl.DateTimeFormatOptions = {
      year: "numeric",
      month: "long",
      day: "numeric",
    };
    return newDate.toLocaleDateString("en-US", options);
  };

  return (
    <div id="background-container">
      <Grid>
        <Grid.Row>
          <Grid.Column widescreen={4}>
            {skiAreaState.loading ? (
              <Dimmer active>
                <Loader>Loading...</Loader>
              </Dimmer>
            ) : skiAreas && skiAreas.length > 0 ? (
              <Card id="ski-area-card">
                {" "}
                {/* max height overflow auto */}
                {skiAreas.map((sa) => (
                  <Card.Content key={sa.slug}>
                    <Link to={`/ski-areas/${sa.slug}`}>{sa.name}</Link>
                  </Card.Content>
                ))}
              </Card>
            ) : skiAreaState.error ? (
              <Dimmer active>
                <Header as="h1">
                  Error! Ski Area Data cannot be retrieved! {skiAreaState.error}
                </Header>
              </Dimmer>
            ) : null}
          </Grid.Column>
          <Grid.Column width={8}>
            <Segment>{isAuthenticated ? <MyPage /> : <AnonHome />}</Segment>
          </Grid.Column>
          <Grid.Column width={4}>
            {reviewState.loading ? (
              <Dimmer active>
                <Loader>Loading...</Loader>
              </Dimmer>
            ) : reviewState.error ? (
              <Dimmer active>
                <Header as="h1">
                  Error! Review Data cannot be retrieved! {reviewState.error}
                </Header>
              </Dimmer>
            ) : isAuthenticated && username ? (
              <div id="inbox-div">
                <Inbox username={username} />
              </div>
            ) : reviews && reviews.length > 0 ? (
              <div>
                <Grid.Row style={{ width: "250px" }}>
                  {reviews.map((review) => (
                    <Grid.Column>
                      {" "}
                      {/* paginated */}
                      <Segment key={review.id} id="review-card">
                        <Header as="h3">
                          <Link to={`/ski-areas/reviews/${review.id}`}>
                            {review.header}
                          </Link>{" "}
                        </Header>
                        <Header as="h4">{review.skiAreaName}</Header>
                        <p>By {review.username}</p>
                        <Divider />
                        <Rating
                          icon="star"
                          defaultRating={review.stars}
                          maxRating={5}
                          disabled
                        />
                        <Divider />
                        {formatDate(review.createdAt)}
                      </Segment>
                    </Grid.Column>
                  ))}
                </Grid.Row>
                <Pagination
                  activePage={activePage}
                  totalPages={totalPages}
                  onPageChange={handlePaginationChange}
                />
              </div>
            ) : null}
          </Grid.Column>
        </Grid.Row>
      </Grid>
    </div>
  );
};

export default Home;
