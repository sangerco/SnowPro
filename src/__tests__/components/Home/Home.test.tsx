import React from "react";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import nock from "nock";
import { Provider } from "react-redux";
import { store } from "../../../redux/store";
import { BrowserRouter } from "react-router-dom";

import Home from "../../../components/Home/Home";

describe("test home component", () => {
  it("should render", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Home />
        </BrowserRouter>
      </Provider>
    );
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Home />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders loader", async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Home />
        </BrowserRouter>
      </Provider>
    );

    const loader = screen.getByTestId("loader");
    expect(loader).toBeInTheDocument();
  });
  it("should render ski area data and review data", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/ski-areas")
      .reply(200, [
        {
          slug: "vail",
          name: "Vail",
        },
        {
          slug: "arapahoe-basin",
          name: "Arapahoe Basin",
        },
      ]);
    nock("http://localhost:5000")
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/ski-areas/reviews")
      .reply(200, [
        {
          id: "review-1",
          header: "test review",
          skiAreaName: "test ski area",
          username: "testuser",
          stars: 5,
          createdAt: "12/11/22 12:22:22.00",
        },
        {
          id: "review-2",
          header: "test review 2",
          skiAreaName: "test ski area 2",
          username: "testuser2",
          stars: 1,
          createdAt: "12/12/22 12:22:22.00",
        },
      ]);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Home />
        </BrowserRouter>
      </Provider>
    );

    await waitFor(() => {
      const skiAreaCard = screen.getByTestId("ski-area-card");
      expect(skiAreaCard).toBeInTheDocument();
    });

    await waitFor(() => {
      const reviewCard = screen.getByTestId("review-card");
      expect(reviewCard).toBeInTheDocument();
    });
  });
});
