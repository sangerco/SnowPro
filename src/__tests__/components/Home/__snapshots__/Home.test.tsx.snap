// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test home component matches snapshot 1`] = `
<DocumentFragment>
  <div
    id="background-container"
  >
    <div
      class="ui grid"
    >
      <div
        class="row"
      >
        <div
          class="four wide widescreen column"
        >
          <div
            class="ui active transition visible dimmer"
            style="display: flex;"
          >
            <div
              class="content"
            >
              <div
                class="ui text loader"
                data-testid="loader"
              >
                Loading...
              </div>
            </div>
          </div>
        </div>
        <div
          class="eight wide column"
        >
          <div
            class="ui segment"
          >
            <div
              class="ui text container"
            >
              <h1
                class="ui green header"
              >
                Welcome to SnowPro
              </h1>
              <h2
                class="ui green header"
              >
                Your hub for ski conditions, reviews and more!
              </h2>
              <div
                class="ui divider"
              />
              <div
                class="ui raised segment"
              >
                <h4
                  class="ui header"
                >
                  <a
                    data-testid="loginLink"
                    href="/login"
                  >
                    Log In
                  </a>
                </h4>
                <div
                  class="ui divider"
                />
                <h4
                  class="ui header"
                >
                  <a
                    data-testid="registerLink"
                    href="/register"
                  >
                    Sign Up
                  </a>
                </h4>
              </div>
              <div
                class="ui divider"
              />
              <div
                class="ui raised segment"
              >
                <h4
                  class="ui header"
                >
                  <a
                    data-testid="skiAreasLink"
                    href="/ski-areas"
                  >
                    Find Ski Areas
                  </a>
                </h4>
              </div>
            </div>
          </div>
        </div>
        <div
          class="four wide column"
        >
          <div
            class="ui active transition visible dimmer"
            style="display: flex;"
          >
            <div
              class="content"
            >
              <div
                class="ui text loader"
              >
                Loading...
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
